### PR TITLE
Add lengthof operator for string and blob

### LINKED ISSUE #4425: Support 'lengthof' operator on string and blob
**Description:**
The `string` and `blob` types represent sequences. Therefore It makes sense to get the length of the sequence using the `lengthof`  unary operator.  

This operator is already supported for the `xml` type.

### LINKED ISSUE #4485: Use of template literals causes internal error
**Description:**
Below program exits with internal error.
```ballerina
function main (string[] args) {
	string name = "John";
	println(string `Hello {{name}}`);
}
```
This is caused by using string template literal as a RHS expression.

**Steps to reproduce:**
See above

**Affected Versions:**
0.95.8-SNAPSHOT

### PR DESCRIPTION
**Pull request by SupunS**
This PR  adds the support for lengthof operator for string and blob types (Resolve #4425).

Also fixes the issues with string template literals (Fix #4485).

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java [hunk 1]
```diff
@@ -415,14 +415,26 @@ private void exec() {
...
-                    j = operands[1];
-                    if (sf.refRegs[i] == null) {
-                        handleNullRefError();
+                    cpIndex = operands[1];
+                    j = operands[2];
+
+                    typeRefCPEntry = (TypeRefCPEntry) constPool[cpIndex];
+                    int typeTag = typeRefCPEntry.getType().getTag();
+                    if (typeTag == TypeTags.STRING_TAG) {
+                        sf.longRegs[j] = sf.stringRegs[i].length();
```
Comment (hasithaa) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java]: this can return a null pointer exception.

Comment (SupunS) [reply modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java]: Fixed

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java [hunk 2]
```diff
@@ -415,14 +415,32 @@ private void exec() {
 
                 case InstructionCodes.LENGTHOF:
                     i = operands[0];
-                    j = operands[1];
-                    if (sf.refRegs[i] == null) {
-                        handleNullRefError();
+                    cpIndex = operands[1];
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java]: Please refactor this instruction code to a new method.

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java [hunk 3]
```diff
@@ -2762,8 +2757,9 @@ public void visit(BLangXMLQuotedString xmlQuotedString) {
 
     @Override
     public void visit(BLangStringTemplateLiteral stringTemplateLiteral) {
-        stringTemplateLiteral.concatExpr.regIndex = createLHSRegIndex(stringTemplateLiteral.regIndex);
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: Can you explain why you did this change?

Comment (SupunS) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: This is to fix https://github.com/ballerinalang/ballerina/issues/4485.
In the prev impl, at this point,  `stringTemplateLiteral.regIndex` is not set, hence inside `createLHSRegIndex()` a null pointer is thrown.  So modeled this in the same way as \[1\]

\[1\] https://github.com/ballerinalang/ballerina/pull/4489/files/e28052fdf6c84eea1876984337ab0ad4e9fe722a#diff-5e3b169d288a55d67a94efc766adbc17L2757

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java (hunk 2 above)
Comment (SupunS) [reply modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java]: Refactored.

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java (hunk 3 above)
Comment (sameera-jayasoma) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: Thanks for the explanation.  I hope you've added a test case.

Comment (SupunS) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: Yes, its included in https://github.com/ballerinalang/ballerina/pull/4489/files/e28052fdf6c84eea1876984337ab0ad4e9fe722a#diff-fe62d65b5ce555bebf6272d23f30aa08R159

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java [hunk 4]
```diff
@@ -3787,4 +3762,51 @@ private AttributeInfo getAttributeInfo(AttributeInfoPool attrInfoPool, Attribute
     private boolean isWaitingOnNonBlockingAction() {
         return context.nonBlockingContext != null;
     }
+
+    private void calculateLength(int[] operands, StackFrame sf) {
+        int i = operands[0];
+        int cpIndex = operands[1];
+        int j = operands[2];
+
+        TypeRefCPEntry typeRefCPEntry = (TypeRefCPEntry) constPool[cpIndex];
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java]: Since there are 4 different types, I would prefer a switch case statement. You can do that as an improvement or another PR

Comment (SupunS) [reply modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java]: Yah, I also thought about it. But then we have to do the null check for each case, for ref-types.. Now we do the null check for all ref-types only once.
WDYT?
